lib/pure/decimal.nim

```diff
@@ -0,0 +1,1040 @@
+      if DIGITS.contains(ch):
+        state = psFracCoeff
+      else:
+        state = psDone
+    of psFracCoeff:
+      if DIGITS.contains(ch):
+        discard
+      elif IGNORED_CHARS.contains(ch):
+        discard
+      elif ch == 'e':
+        state = psSignForExp
+      else:
+        state = psDone
+    of psSignForExp:
+      if DIGITS.contains(ch):
+        state = psExp
+      elif (ch == '-') or (ch == '+'):
+        discard
+      else:
+        state = psDone
+    of psExp:
+      if DIGITS.contains(ch):
+        discard
+      else:
+        state = psDone
+    of psDone:
+      discard
+    #
+    # act on state
+    #
+    case state:
+    of psPre:
+      discard
+    of psLeadingMinus:
+      negative = true
+    of psLeadingZeroes:
+      legit = true
+    of psIntCoeff:
+      # given the state table, the 'find' function should never return -1
+      if not IGNORED_CHARS.contains(ch):
+        digitList.add(DIGITS.find(ch).byte)
+        legit = true
+    of psDecimalPoint:
+      discard
+    of psFracCoeff:
+      # given the state table, the 'find' function should never return -1
+      if not IGNORED_CHARS.contains(ch):
+        digitList.add(DIGITS.find(ch).byte)
+        exponent -= 1
+        legit = true
+    of psSignForExp:
+      if ch == '-':
+        expNegative = true
+    of psExp:
+      expDigitList &= ch
+    of psDone:
+      discard
+  #
+  # remove leading zeroes
+  #
+  var nonZeroFound = false
+  var temp: seq[byte] = @[]
+  for val in digitList:
+    if val != 0.byte:
+      nonZeroFound = true
+    if nonZeroFound:
+      temp.add val
+  digitList = temp
+  #
+  # if too many digits, removing trailing digits.
+  # Note: because this is on 70-digit Transient, simple "truncation" is good enough
+  #
+  if digitList.len > tempSignificandSize:
+    let digitsToRemove = digitList.len - tempSignificandSize
+    digitList = digitList[0 ..< tempSignificandSize]
+    exponent += digitsToRemove
+  #
+  # move to result to final significand
+  #
+  let offset = tempSignificandSize - digitList.len
+  for index, val in digitList:
+    significand[index + offset] = val
+  #
+  # parse the exponent value
+  #
+  if expDigitList.len > 0:
+    try:
+      let exp = parseInt(expDigitList)
+      if expNegative:
+        exponent -= exp
+      else:
+        exponent += exp
+      if exponent < expLowerBound:
+        let shiftNeeded = (expLowerBound - exponent).int16
+        significand = shiftDecimalsRightTransient(significand, shiftNeeded)
+        exponent += shiftNeeded
+    except:
+      discard
+  #
+  result = (dkValued, negative, significand, exponent)
+
+#
+# attributes
+#
+
+proc negative*(number: Decimal): bool =
+  when defined(js):
+    if (number.a and signMask) shr 0 == signMask:
+      result = true
+    else:
+      result = false
+  else:
+    if (number.a and signMask) == signMask:
+      result = true
+    else:
+      result = false
+
+proc significance*(number: Decimal): int =
```

Comment:
```suggestion
proc significants*(number: Decimal): int =
```
I haven't heard "significance" being used for that yet.

Reply:
https://en.wikipedia.org/wiki/Significance_arithmetic

Still, if you want, I could use something more explicit like `significantDigitCount` or `qtyOfSignificantDigits`. When I see `significants` it implies we are wanting a list of the significant digits; at least to me.

Reply:
I prefer `significantDigitCount` too.

Reply:
next commit changes this to `significantDigitCount`